feat(zk): implement halo2 permutation keygen

dongchangYoo opened the pull request:
This PR implements `keygen` logics for copy permutation. (see, [Permutation Assembly in Halo2](https://github.com/kroma-network/halo2/blob/dev/halo2_proofs/src/plonk/permutation/keygen.rs))

Once the `KZGParams` class is completed, an update will be made to add the `commit()` logic to `PermutationAssembly`.

chokobole commented on tachyon/zk/plonk/permutation/permutation_keys.h:
```diff
@@ -0,0 +1,117 @@
+template <typename Curve>
+class PermutationVerifyingKey {
+ public:
+  using CommitmentTy = math::AffinePoint<Curve>;
+  using CommitmentVec = std::vector<CommitmentTy>;
```
```suggestion
  using Commitment = math::AffinePoint<Curve>;
  using Commitments = std::vector<CommitmentTy>;
```

chokobole replied:
Or I think you can just use the type directly without alias since the alias type is not that shorter.

dongchangYoo replied:
`AffinePoint` used an alias to indicate that it is a `Commitment`. And I will change the name of the alias to something like suggestion.

chokobole commented on tachyon/zk/plonk/permutation/permutation_keys.h:
```diff
@@ -0,0 +1,117 @@
+                        std::vector<DensePoly>&& polys)
+      : permutations_(std::move(permutations)), polys_(std::move(polys)) {}
+
+  size_t BytesLength() const { NOTIMPLEMENTED(); };
```
Why is that?

dongchangYoo replied:
it is missing implementation.. i will resolve this!

chokobole commented on tachyon/zk/plonk/permutation/permutation_keys.h:
```diff
@@ -0,0 +1,117 @@
+
+  size_t BytesLength() const { NOTIMPLEMENTED(); };
+
+  const std::vector<Evals>& permutation() const { return permutations_; }
+  const std::vector<DensePoly>& poly() const { return polys_; }
```
```suggestion
  const std::vector<Evals>& permutations() const { return permutations_; }
  const std::vector<DensePoly>& polys() const { return polys_; }
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_keys.h:
```diff
@@ -0,0 +1,117 @@
+
+  static bool ReadFrom(const Buffer& buffer,
+                       zk::PermutationVerifyingKey<Curve>* vk) {
+    using VerifyingKey = zk::PermutationVerifyingKey<Curve>;
+
+    typename VerifyingKey::CommitmentVec commitments;
+    if (!buffer.Read(&commitments)) return false;
+    *vk = VerifyingKey(std::move(commitments));
+    return true;
```
```suggestion
    std::vector<math::AffinePoint<Curve>> commitments;
    if (!buffer.Read(&commitments)) return false;
    *vk = zk::PermutationVerifyingKey<Curve>(std::move(commitments));
    return true;
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_keys.h:
```diff
@@ -0,0 +1,117 @@
+    math::UnivariateDensePolynomial<F, MaxDegree> perms;
+    math::UnivariateEvaluations<F, MaxDegree> poly;
+    if (!buffer.ReadMany(&perms, &poly)) return false;
+
+    return zk::PermutationProvingKey<F, MaxDegree>(std::move(perms),
+                                                   std::move(poly));
```
```suggestion
    *pk = zk::PermutationProvingKey<F, MaxDegree>(std::move(perms), std::move(poly));
    return true;                                                   
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_keys.h:
```diff
@@ -0,0 +1,117 @@
+namespace tachyon {
+namespace zk {
+
+template <typename Curve>
+class PermutationVerifyingKey;
```
You don't  need this.

chokobole commented on tachyon/zk/plonk/permutation/permutation_keys.h:
```diff
@@ -0,0 +1,117 @@
+#define TACHYON_ZK_PLONK_PERMUTATION_KEYS_H_
+
+#include "tachyon/base/buffer/copyable.h"
+#include "tachyon/export.h"
```
Please remove this.

chokobole commented on tachyon/zk/plonk/permutation/permutation_argument.h:
```diff
@@ -0,0 +1,39 @@
+
+namespace tachyon::zk {
+
+// TODO(dongchangYoo): This class just provides shape of `PermutationArgument`.
```
```suggestion
// TODO(dongchangYoo): This class just provides shape of |PermutationArgument|.
```

chokobole commented on tachyon/zk/plonk/permutation/BUILD.bazel:
```diff
@@ -11,3 +11,27 @@ tachyon_cc_library(
+    name = "permutation_argument",
+    hdrs = ["permutation_argument.h"],
+    deps = [
+        ":permutation_keys",
```
Please remove this since you don't use this.

chokobole commented on tachyon/zk/plonk/permutation/BUILD.bazel:
```diff
@@ -11,3 +11,27 @@ tachyon_cc_library(
+        "//tachyon/base:logging",
+        "//tachyon/base/containers:contains",
+        "//tachyon/zk/plonk/circuit:column",
+        "@com_google_googletest//:gtest_prod",
```
```suggestion
        "//tachyon/math/polynomials/univariate:univariate_polynomial",
        "@com_google_googletest//:gtest_prod",
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_argument.h:
```diff
@@ -0,0 +1,39 @@
+    columns_.push_back(column);
+  }
+
+  const std::vector<AnyColumn>& columns() const { return columns_; }
+  std::vector<AnyColumn>& columns() { return columns_; }
```
Please move getters above.

chokobole commented on tachyon/zk/plonk/permutation/permutation_argument.h:
```diff
@@ -0,0 +1,39 @@
+class PermutationArgument {
+ public:
+  PermutationArgument() = default;
+  PermutationArgument(const std::vector<AnyColumn>& columns)
```
```suggestion
  explicit PermutationArgument(const std::vector<AnyColumn>& columns)
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_argument_unittest.cc:
```diff
@@ -0,0 +1,26 @@
+#include "gtest/gtest.h"
+
+#include "tachyon/math/elliptic_curves/short_weierstrass/test/curve_config.h"
+#include "tachyon/math/finite_fields/test/gf7.h"
```
You don't need this.

chokobole commented on tachyon/zk/plonk/permutation/permutation_argument_unittest.cc:
```diff
@@ -0,0 +1,26 @@
+TEST(PermutationArgument, Constructor) {
+  using CurveTy = math::test::AffinePoint::Curve;
+
+  const AnyColumn col1 = AnyColumn(0);
+  const AnyColumn col2 = AdviceColumn(1);
+  const AnyColumn col3 = FixedColumn(2);
+  std::vector<AnyColumn> columns = {col1, col2, col3};
```
```suggestion
  std::vector<AnyColumn> columns = {AnyColumn(0), AnyColumn(1), AnyColumn(2)};
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_argument_unittest.cc:
```diff
@@ -0,0 +1,26 @@
+
+namespace tachyon::zk {
+
+TEST(PermutationArgument, Constructor) {
```
```suggestion
TEST(PermutationArgumentTest, Constructor) {
```

chokobole commented on tachyon/zk/plonk/permutation/BUILD.bazel:
```diff
@@ -24,14 +24,30 @@ tachyon_cc_library(
+    name = "permutation_assembly",
+    hdrs = ["permutation_assembly.h"],
+    deps = [
+        ":permutation_argument",
```
```suggestion
        ":permutation_argument",
        ":permutation_keys",
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_assembly.h:
```diff
@@ -0,0 +1,245 @@
+template <typename Curve, size_t MaxDegree>
+class PermutationAssembly {
+ public:
+  struct Label {
+    size_t col = 0;
+    size_t row = 0;
```
Please make it class.

chokobole commented on tachyon/zk/plonk/permutation/BUILD.bazel:
```diff
@@ -24,14 +24,30 @@ tachyon_cc_library(
+        ":permutation_argument",
+        "//tachyon/base:openmp_util",
+        "//tachyon/base/containers:container_util",
+        "//tachyon/math/elliptic_curves:points",
```
Please remove this.

chokobole commented on tachyon/zk/plonk/permutation/BUILD.bazel:
```diff
@@ -24,14 +24,30 @@ tachyon_cc_library(
+        "//tachyon/base/containers:container_util",
+        "//tachyon/math/elliptic_curves:points",
+        "//tachyon/math/polynomials/univariate:univariate_evaluation_domain",
+        "@com_google_googletest//:gtest_prod",
```
Please remove this.

chokobole commented on tachyon/zk/plonk/permutation/permutation_assembly.h:
```diff
@@ -0,0 +1,245 @@
+  using Mapping = std::vector<std::vector<Label>>;
+
+  using CommitmentTy = math::AffinePoint<Curve>;
+  using FieldTy = typename Curve::ScalarField;
```
```suggestion
  using ScalarField = typename Curve::ScalarField;
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_assembly.h:
```diff
@@ -0,0 +1,245 @@
+
+  using EvaluationDomainTy =
+      typename math::UnivariateEvaluationDomain<FieldTy, MaxDegree>;
+  using EvalsPolyTy = math::UnivariateEvaluations<FieldTy, MaxDegree>;
+  using CoeffsPolyTy = math::UnivariateDensePolynomial<FieldTy, MaxDegree>;
```
```suggestion
  using Evals = math::UnivariateEvaluations<FieldTy, MaxDegree>;
  using DensePoly = math::UnivariateDensePolynomial<FieldTy, MaxDegree>;
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_assembly_unittest.cc:
```diff
@@ -0,0 +1,141 @@
+      assembly_.BuildVerifyingKey(domain.get());
+  PermutationProvingKey<F, MaxDegree> pk_ori =
+      assembly_.BuildProvingKey(domain.get());
+  // PermutationVerifyingKey<Curve> vk;
+  // PermutationProvingKey<F, MaxDegree> pk;
+
+  // base::VectorBuffer buf;
+  // buf.WriteMany(vk_ori, pk_ori);
+
+  // buf.set_buffer_offset(0);
+  // buf.ReadMany(&vk, &pk);
```
What are these?, what are you going to test?

chokobole commented on tachyon/zk/plonk/permutation/permutation_assembly_unittest.cc:
```diff
@@ -0,0 +1,141 @@
+using AsmLabel = typename AsmTy::Label;
+using EvalsPolyTy = typename PermutationAssembly<Curve, MaxDegree>::EvalsPolyTy;
+
+std::vector<AnyColumn> COLUMNS = {AnyColumn(0), AdviceColumn(1), FixedColumn(2),
+                                  InstanceColumn(3)};
```
Do not declare `std::vector<>` global variable.

Seee https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

chokobole commented on tachyon/zk/plonk/permutation/permutation_assembly.h:
```diff
@@ -0,0 +1,245 @@
+#if defined(TACHYON_HAS_OPENMP)
+    size_t thread_num = omp_get_max_threads();
+#else
+    int thread_num = 1;
```
```suggestion
    size_t thread_num = 1;
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_assembly.h:
```diff
@@ -0,0 +1,245 @@
+
+  // Generate the permutation polynomials based on the accumulated copy
+  // permutations. Note that the permutation polynomials are in evaluation form.
+  constexpr std::vector<EvalsPolyTy> GeneratePermutations(
```
Can't be run at compile time. It's multithreaded.

```suggestion
  std::vector<EvalsPolyTy> GeneratePermutations(
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_assembly.h:
```diff
@@ -0,0 +1,245 @@
+
+  PermutationAssembly() = default;
+
+  // Constructor with `PermutationArgument` which has copy permutation columns.
+  PermutationAssembly(const PermutationArgument<Curve, MaxDegree>& p)
```
```suggestion
  // Constructor with `PermutationArgument` which has copy permutation columns.
  explicit PermutationAssembly(const PermutationArgument<Curve, MaxDegree>& p)
```

TomTaehoonKim commented:
@dongchangYoo
1. Could you please add the perma links to the commit f6c080c, 091cdbc, c130038 as well?
2. Can you fix the typo in the description of c61e341? PermutationArguemtn -> PermutationArgument.

TomTaehoonKim commented on tachyon/zk/plonk/permutation/permutation_assembly.h:
```diff
@@ -0,0 +1,186 @@
+        //     columns.size(), base::CreateVector(MaxDegree + 1, size_t{1}))) {}
+        sizes_(Mapping<size_t>::Create(columns.size(), MaxDegree + 1,
+                                       size_t(1))) {}
+  const Mapping<Label>& mapping() const { return mapping_; }
```
```suggestion

  const Mapping<Label>& mapping() const { return mapping_; }
```
Can you add an empty line?

dongchangYoo commented:
`Commitment` is a type alias, and `LookupTable` and `Label` are wrapper classes for smoothly using two-dimensional vectors, so there are no relevant links for them. 
I've added the other links. Also, I've corrected the link-cut and the typo.

dongchangYoo commented:
> @dongchangYoo
> 
> 1. Could you please add the perma links to the commit [f6c080c](https://github.com/kroma-network/tachyon/commit/f6c080c505e58eecf0582c8aca0408796c0e28fc), [091cdbc](https://github.com/kroma-network/tachyon/commit/091cdbc78eb25ba062a5d65615c1acc3dbb4e5fd), [c130038](https://github.com/kroma-network/tachyon/commit/c130038291856fb3bcf8fa8bc0828ad15f9d50ee) as well?
> 2. Can you fix the typo in the description of [c61e341](https://github.com/kroma-network/tachyon/commit/c61e341e615297a6bd39c84014f7511e70e34e77)? PermutationArguemtn -> PermutationArgument.
> 3. Can you fix the link in the description of [c61e341](https://github.com/kroma-network/tachyon/commit/c61e341e615297a6bd39c84014f7511e70e34e77)? I think kind of cut at the front.
> 4. Can you fix the link in the descript of [2df1902](https://github.com/kroma-network/tachyon/commit/2df1902c2cd0dc31b8856df6f4deffa7bd347153) as well?

Commitment is a type alias, and LookupTable and Label are wrapper classes for smoothly using two-dimensional vectors, so there are no relevant links for them.
I've added the other links. Also, I've corrected the link-cut and the typo.

TomTaehoonKim commented on tachyon/zk/plonk/permutation/lookup_table.h:
```diff
@@ -0,0 +1,85 @@
+#include <utility>
+#include <vector>
+
+#include "gtest/gtest_prod.h"
```
@dongchangYoo Oh, we need it for `FRIEND_TEST` never mind.

TomTaehoonKim commented on tachyon/zk/plonk/permutation/lookup_table.h:
```diff
@@ -0,0 +1,85 @@
+
+  static LookupTable Construct(
+      const size_t size,
+      math::UnivariateEvaluationDomain<F, MaxDegree>* domain) {
```
@dongchangYoo Can you fix this please? See https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs
> Non-optional input parameters should usually be values or const references

TomTaehoonKim commented on tachyon/zk/plonk/permutation/permutation_argument_unittest.cc:
```diff
@@ -0,0 +1,33 @@
+  columns.push_back(col2);
+  EXPECT_EQ(argument.columns(), columns);
+
+  // Newly Instance column should be added.
```
@dongchangYoo Please check this once more please!

TomTaehoonKim commented on tachyon/zk/plonk/permutation/permutation_assembly.h:
```diff
@@ -0,0 +1,186 @@
+
+  PermutationAssembly() = default;
+
+  // Constructor with |PermutationArgument| which has copy permutation columns.
```
@dongchangYoo Please check this once more please!

chokobole commented on tachyon/zk/plonk/permutation/permutation_proving_key.h:
```diff
@@ -0,0 +1,80 @@
+
+  PermutationProvingKey() = default;
+
+  explicit PermutationProvingKey(const std::vector<Evals>& permutations,
+                                 const std::vector<DensePoly>& polys)
+      : permutations_(permutations), polys_(polys) {}
+
+  explicit PermutationProvingKey(std::vector<Evals>&& permutations,
+                                 std::vector<DensePoly>&& polys)
+      : permutations_(std::move(permutations)), polys_(std::move(polys)) {}
```
```suggestion
  PermutationProvingKey(const std::vector<Evals>& permutations,
                                 const std::vector<DensePoly>& polys)
      : permutations_(permutations), polys_(polys) {}

  PermutationProvingKey(std::vector<Evals>&& permutations,
                                 std::vector<DensePoly>&& polys)
      : permutations_(std::move(permutations)), polys_(std::move(polys)) {}
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_proving_key_unittest.cc:
```diff
@@ -0,0 +1,40 @@
+
```
```suggestion
// Copyright 2020-2022 The Electric Coin Company
// Copyright 2022 The Halo2 developers
// Use of this source code is governed by a MIT/Apache-2.0 style license that
// can be found in the LICENSE-MIT.halo2 and the LICENCE-APACHE.halo2
// file.

```

chokobole commented on tachyon/zk/plonk/permutation/permutation_verifying_key_unittest.cc:
```diff
@@ -0,0 +1,35 @@
+
```
```suggestion
// Copyright 2020-2022 The Electric Coin Company
// Copyright 2022 The Halo2 developers
// Use of this source code is governed by a MIT/Apache-2.0 style license that
// can be found in the LICENSE-MIT.halo2 and the LICENCE-APACHE.halo2
// file.

```

chokobole commented on tachyon/zk/plonk/permutation/lookup_table_unittest.cc:
```diff
@@ -0,0 +1,46 @@
+
```
```suggestion
// Copyright 2020-2022 The Electric Coin Company
// Copyright 2022 The Halo2 developers
// Use of this source code is governed by a MIT/Apache-2.0 style license that
// can be found in the LICENSE-MIT.halo2 and the LICENCE-APACHE.halo2
// file.
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_proving_key.h:
```diff
@@ -0,0 +1,80 @@
+#ifndef TACHYON_ZK_PLONK_PERMUTATION_PERMUTATION_PROVING_KEY_H_
```
```suggestion
// Copyright 2020-2022 The Electric Coin Company
// Copyright 2022 The Halo2 developers
// Use of this source code is governed by a MIT/Apache-2.0 style license that
// can be found in the LICENSE-MIT.halo2 and the LICENCE-APACHE.halo2
// file.

#ifndef TACHYON_ZK_PLONK_PERMUTATION_PERMUTATION_PROVING_KEY_H_
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_verifying_key.h:
```diff
@@ -0,0 +1,67 @@
+#ifndef TACHYON_ZK_PLONK_PERMUTATION_PERMUTATION_VERIFYING_KEY_H_
```
```suggestion
// Copyright 2020-2022 The Electric Coin Company
// Copyright 2022 The Halo2 developers
// Use of this source code is governed by a MIT/Apache-2.0 style license that
// can be found in the LICENSE-MIT.halo2 and the LICENCE-APACHE.halo2
// file.

#ifndef TACHYON_ZK_PLONK_PERMUTATION_PERMUTATION_VERIFYING_KEY_H_
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_argument.h:
```diff
@@ -0,0 +1,31 @@
+#ifndef TACHYON_ZK_PLONK_PERMUTATION_PERMUTATION_ARGUMENT_H_
```
```suggestion
// Copyright 2020-2022 The Electric Coin Company
// Copyright 2022 The Halo2 developers
// Use of this source code is governed by a MIT/Apache-2.0 style license that
// can be found in the LICENSE-MIT.halo2 and the LICENCE-APACHE.halo2
// file.

#ifndef TACHYON_ZK_PLONK_PERMUTATION_PERMUTATION_ARGUMENT_H_
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_argument.h:
```diff
@@ -0,0 +1,31 @@
+class PermutationArgument {
+ public:
+  PermutationArgument() = default;
+  explicit PermutationArgument(const std::vector<AnyColumn>& columns)
+      : columns_(columns) {}
```
```suggestion
  explicit PermutationArgument(const std::vector<AnyColumn>& columns)
      : columns_(columns) {}
  explicit PermutationArgument(std::vector<AnyColumn>&& columns)
      : columns_(std::move(columns)) {}
```

chokobole commented on tachyon/zk/plonk/permutation/cycle_store_unittest.cc:
```diff
@@ -0,0 +1,48 @@
+#include "tachyon/zk/plonk/permutation/cycle_store.h"
```
```suggestion
// Copyright 2020-2022 The Electric Coin Company
// Copyright 2022 The Halo2 developers
// Use of this source code is governed by a MIT/Apache-2.0 style license that
// can be found in the LICENSE-MIT.halo2 and the LICENCE-APACHE.halo2
// file.

#include "tachyon/zk/plonk/permutation/cycle_store.h"
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_assembly.h:
```diff
@@ -0,0 +1,151 @@
+  // Constructor with permutation columns.
+  explicit PermutationAssembly(const std::vector<AnyColumn>& columns)
+      : columns_(columns), cycle_store_(CycleStore(columns.size(), kRows)) {}
+
```
```suggestion
  // Constructor with permutation columns.
  explicit PermutationAssembly(std::vector<AnyColumn>&& columns)
      : columns_(std::move(columns)), cycle_store_(CycleStore(columns_.size(), kRows)) {}
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_assembly.h:
```diff
@@ -0,0 +1,151 @@
+
+  // Constructor with permutation columns.
+  explicit PermutationAssembly(const std::vector<AnyColumn>& columns)
+      : columns_(columns), cycle_store_(CycleStore(columns.size(), kRows)) {}
```
For consistency with below,

```suggestion
      : columns_(columns), cycle_store_(CycleStore(columns_.size(), kRows)) {}
```

chokobole commented on tachyon/zk/plonk/permutation/permutation_assembly_unittest.cc:
```diff
@@ -0,0 +1,65 @@
+      LookupTable<F, MaxDegree>::Construct(columns_.size(), domain_.get());
+
+  for (size_t i = 0; i < columns_.size(); ++i) {
+    std::vector<F> permutation = permutations[i].evaluations();
```
Considering `permutation` is an element of `permutations[i]`, should it be `evaluations`?
```suggestion
    const std::vector<F>& evaluations = permutations[i].evaluations();
```

TomTaehoonKim commented on tachyon/zk/plonk/permutation/permutation_proving_key.h:
```diff
@@ -0,0 +1,80 @@
+  using DensePoly = math::UnivariateDensePolynomial<F, MaxDegree>;
+  using Evals = math::UnivariateEvaluations<F, MaxDegree>;
+
+  PermutationProvingKey() = default;
+
+  explicit PermutationProvingKey(const std::vector<Evals>& permutations,
+                                 const std::vector<DensePoly>& polys)
+      : permutations_(permutations), polys_(polys) {}
+
+  explicit PermutationProvingKey(std::vector<Evals>&& permutations,
```
```suggestion
  PermutationProvingKey() = default;
  explicit PermutationProvingKey(const std::vector<Evals>& permutations,
                                 const std::vector<DensePoly>& polys)
      : permutations_(permutations), polys_(polys) {}
  explicit PermutationProvingKey(std::vector<Evals>&& permutations,
```

chokobole replied:
I was thinking is it good to say about it, and you did LoL

TomTaehoonKim commented on tachyon/zk/plonk/permutation/permutation_verifying_key.h:
```diff
@@ -0,0 +1,67 @@
+template <typename Curve>
+class PermutationVerifyingKey {
+ public:
+  PermutationVerifyingKey() = default;
+
+  explicit PermutationVerifyingKey(const Commitments<Curve>& commitments)
+      : commitments_(commitments) {}
+
+  explicit PermutationVerifyingKey(Commitments<Curve>&& commitments)
+      : commitments_(std::move(commitments)) {}
```
Can. you remove the empty lines?
```suggestion
  PermutationVerifyingKey() = default;
  explicit PermutationVerifyingKey(const Commitments<Curve>& commitments)
      : commitments_(commitments) {}
  explicit PermutationVerifyingKey(Commitments<Curve>&& commitments)
      : commitments_(std::move(commitments)) {}
```

chokobole reviewed (approved):
LGTM

TomTaehoonKim reviewed (approved):
LGTM! Great work!
